trieste/acquisition/function/greedy_batch.py

```diff
@@ -351,3 +355,286 @@ def __call__(self, x: TensorType) -> TensorType:
+            base_acquisition_function_builder = ExpectedImprovement()
+
+        if isinstance(base_acquisition_function_builder, SingleModelAcquisitionBuilder):
+            base_acquisition_function_builder = base_acquisition_function_builder.using(OBJECTIVE)
+
+        self._builder = base_acquisition_function_builder
+        self._fantasize_method = fantasize_method
+
+    def prepare_acquisition_function(
+        self,
+        models: Mapping[str, ProbabilisticModel],
+        datasets: Optional[Mapping[str, Dataset]] = None,
+        pending_points: Optional[TensorType] = None,
+    ) -> AcquisitionFunction:
+        """
+
+        :param models: The models over each tag.
+        :param datasets: The data from the observer (optional).
+        :param pending_points: Points already chosen to be in the current batch (of shape [M,D]),
+            where M is the number of pending points and D is the search space dimension.
+        :return: An acquisition function.
+        """
+        if pending_points is None:
+            return self._builder.prepare_acquisition_function(models, datasets)
+        else:
+            tf.debugging.assert_rank(pending_points, 2)
+
+            fantasized_data = {
+                tag: _generate_fantasized_data(
+                    fantasize_method=self._fantasize_method,
+                    model=model,
+                    pending_points=pending_points,
+                )
+                for tag, model in models.items()
+            }
+            new_models = {
+                tag: _fantasize_model(model, fantasized_data[tag]) for tag, model in models.items()
+            }
+
+            if datasets is None:
+                datasets = fantasized_data
+            else:
+                datasets = {tag: data + fantasized_data[tag] for tag, data in datasets.items()}
+
+            return self._builder.prepare_acquisition_function(new_models, datasets)
+
+
+def _generate_fantasized_data(
+    fantasize_method: str, model: ProbabilisticModel, pending_points: TensorType
+) -> Dataset:
+    """
+    Generates "fantasized" data at pending_points depending on the chosen heuristic:
+    - KB (kriging believer) uses the mean prediction of the models
+    - sample uses samples from the GP posterior.
+
+    :param fantasize_method: the following options are available: "KB" and "sample".
+    :param model: a model with predict method
+    :param dataset: past data
+    :param pending_points: points at which to fantasize data
+    :return: a fantasized dataset
+    """
+    if fantasize_method == "KB":
+        fantasized_obs, _ = model.predict(pending_points)
+    elif fantasize_method == "sample":
+        fantasized_obs = model.sample(pending_points, num_samples=1)[0]
+    else:
+        raise NotImplementedError(
+            f"fantasize_method must be KB or sample, " f"received {model.__repr__()}"
+        )
+
+    return Dataset(pending_points, fantasized_obs)
+
+
+def _fantasize_model(
```

Comment:
perhaps a name that is easier to distinguish from _fantasized_model, e.g. _get_fantasized_model?
```suggestion
def _get_fantasized_model(
```